WebAuthn support

jpaniagualaconich opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Implement a new second factor method that uses WebAuthn.

## Description
<!--- Describe your changes in detail -->
A new authentication method is available in the setup wizard: WebAuthn, that enables support for hardware authentication devices such as fido-u2f keys in modern browsers that implement the specification.
 
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
WebAuthn is supported in the Chrome, Firefox, and Edge browsers, for credential creation and assertion using a U2F Token, like those provided by Yubico and Feitian.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
New view and utils tests have been added.
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

moggers87 reviewed (changes requested):
I've only done a quick look over, but here are things that popped out to me right away:

- the new dependency should be optional, like the yubikey dependencies are
- a lot of the translation file updates appear to be unrelated to this PR and make it difficult to see if there's anything other than line number changes in there
- Is the client side JS really required? I'd prefer not to have any JS involved at all if possible. Also there are no tests for this non-trivial code.
- The `TWO_FACTOR_DEVICE_PREFERENCE` stuff looks interesting, but really needs its own PR

moggers87 commented:
By the looks of https://github.com/w3c/webauthn/issues/1255 it seems the JS is not optional, so that's a huge :-1: from me. Not enough for me to reject this PR out of hand, but that code really does need tests and the entirety of webauthn needs to be optional in this package. That JS also needs to be taken out of the templates so people who aren't using the default templates don't have to go write their own JS + tests.

claudep commented:
It would be much easier to host such optional modules if we could finish and merge #331

moggers87 commented:
It's on my todo list!

jpaniagualaconich commented:
Hi there. Shall I wait for #331 in that case? It isn't the best thing in the world that I "if cased" webauthn sprinkling it here and there especially in the SetupView. Mimicking what's been done for the yubikey case to make it optional would only go further down that road.

And no, JS code is not optional. Do you have a preference for the testing framework I should use?

moggers87 commented:
I have no preference on what testing framework you use, whatever you're already familiar with is probably best.

jpaniagualaconich commented:
Hi there!

> * the new dependency should be optional, like the yubikey dependencies are

now it's optional

> * a lot of the translation file updates appear to be unrelated to this PR and make it difficult to see if there's anything other than line number changes in there

took out all the translation-related stuff

> * Is the client side JS really required? I'd prefer not to have any JS involved at all if possible. Also there are no tests for this non-trivial code.

changed it so JS is sent to the client only when WebAuthn is being used. Added an E2E test that covers the whole process (registration and use).
webauthn tests will be skipped if the webauthn package is not installed.

The E2E test will be skipped if:
- one of these packages is missing: webauthn, selenium, webdriver-manager. First one is an extra (`pip install django-two-factor-auth[webauthn]`). For the other two there's a `requirements_e2e.txt` file. 
- the `StaticLiveServerTestCase` service can't be reached at https://localhost/.
I added an `ssl-rev-proxy` make target that sets up the root CA and a localhost certificate needed to serve https.
It also spins an nginx container to reverse proxy from https://localhost.
It needs docker and mkcert.

The "webauthn is an extra" part is in the documentation but the "how to run the E2E test" part is not. I'd like to know what you think of the whole thing before attempting to document it.

> * The `TWO_FACTOR_DEVICE_PREFERENCE` stuff looks interesting, but really needs its own PR

it's no longer a part of this PR.

moggers87 commented on two_factor/templates/two_factor/core/login.html:
```diff
@@ -49,4 +49,35 @@ <h1>{% block title %}{% trans "Login" %}{% endblock %}</h1>
+
+  {{ form.media }}
+  {% if webauthn %}
+    <script>
```
There should be no JS in HTML like this. It makes implementing CSP difficult as you can no longer tell the browser to block inline JS.

jpaniagualaconich replied:
No more inline JS

moggers87 commented on two_factor/templates/two_factor/core/setup.html:
```diff
@@ -53,4 +53,43 @@ <h1>{% block title %}{% trans "Enable Two-Factor Authentication" %}{% endblock %
+  {{ form.media }}
+  
+  {% if wizard.steps.current == 'webauthn' %}
+    <script>
```
As above.

jpaniagualaconich replied:
No more inline JS

moggers87 reviewed (dismissed):
I think we could add the e2e tests to tox.ini in a similar way to how yubikey tests were added.

jpaniagualaconich commented:
- Did away with inline JS 
- No need for docker, mkcert, docker-ngin. django-sslserver is used instead
- Documentation is updated
- Chrome can silently stop cooperating if you forget some settings. Made it harder for this to happen.
- modified tox.ini to include webauthn testing, with E2E bits too

brugnara commented:
Hi @moggers87 

Could you please help us and review this PR?

Thank you for considering my request.

moggers87 commented:
I'd like to wait for #425 to be merged before reviewing any more PRs

jpaniagualaconich commented:
There's a new version of the `py_webauthn` package which brings support for lots of interesting attestation statement formats. Here is the complete list:
https://github.com/duo-labs/py_webauthn/releases/tag/v1.0.0

It does come with a totally new API, though, and now it requires python 3.8+.
@moggers87 since we're waiting for #425 I'm tempted to update. What do you recommend?

Frikster commented:
For those waiting on this who don't want to use this development branch, I have tested and found both the following repos [django-fido](https://github.com/CZ-NIC/django-fido) and [kagi](https://github.com/justinmayer/kagi) to work quite nicely to offer Webauth 2FA.

moggers87 commented:
@jpaniagualaconich not sure I like bumping the minimum required version of Python to 3.8 just yet. 3.7 will still be receiving updates until June 2023: https://endoflife.date/python

jpaniagualaconich commented:
@moggers87 ok, I patched `typing` so it has `Literal` also on python 3.7. This is what the new version of `webauthn` expects and it's the only reason why it wasn't working on 3.7.

I merged latest master and it's now ready for you to review.

moggers87 reviewed (commented):
I got about half way through reviewing this and #331 was ready for merging. That's caused some conflicts.

How much work would it be to move webauthn specific code to the new plugins folder?

claudep commented:
@jpaniagualaconich I'm still working on making this package more plugin-friendly, so some more "breakages" may come soon in master branch. I'm sorry for this, and will try to help as much as possible for your branch integration.

jpaniagualaconich commented:
@claudep ok no problem.

@moggers87 ok, it uses django-extensions now and everything lives in `plugins/webauthn`.

jpaniagualaconich commented:
@moggers87 @claudep I merged latest master and went with the _pluginisation_ as far as the current framework allows.

moggers87 commented:
@claudep I'll do a review, but I wouldn't mind a second pair of eyes to take a look.

claudep commented:
Sure, that's my intention, but I'd like to finish integrating the email plugin first.

hansegucker commented:
Is there any progress on this topic or something one could help with?

jpaniagualaconich commented:
I merged latest master, again. It's ready for someone to review.

claudep commented:
@jpaniagualaconich Do you think you would be able to extract the `backup_phones` replacement by get_devices/get_other_devices in a separate PR? (I may ask you other such separate PR's in the future, hope you don't mind).

claudep commented:
Would it also possible to move the migration file inside plugins/webauthn dir?

jpaniagualaconich commented:
@claudep I extracted the device/other devices thingy into its own PR: https://github.com/jazzband/django-two-factor-auth/pull/526

jpaniagualaconich commented:
@claudep migration file moved, also webauthn tests moved to their own folder (`two_factor/plugins/webauthn/tests/`)
It's ready for another review round.

claudep reviewed (changes requested):
What I'd still like to explore is the option to not push the JS files through the templates engine, but add the webauthn options through the data attributes of the form fields. This could however also be a follow-up optimization, if possible.
Regarding the correctness of the webauthn process, I trust you (and the tests).
Thanks a lot for your work and your patience.

jpaniagualaconich commented on Makefile:
```diff
@@ -1,5 +1,3 @@
-TARGET?=tests
-
```
yes, it is. If left untouched it will only look in the tests folder.

claudep replied:
Ah yes, then it's fine!

jpaniagualaconich commented on .gitignore:
```diff
@@ -8,5 +8,7 @@ example/settings_private.py
+.venv/
 
 .idea/
+.vscode/
```
I sincerely wonder if changes such as this one really get in the way of understanding the rest of the PR.

WhyNotHugo replied:
Hint: Put this kind of ignores in `$HOME/.config/git/ignore`. They'll apply to any repository on which you operate locally, and you won't need to add IDE-specific ignores to every single repo you work on.

claudep replied:
This discussion is now on #534.

For me, having pure self-contained changes with specific commits is important (I didn't say an obsession :smile: )

jpaniagualaconich replied:
excluded from PR as requested

jpaniagualaconich commented on setup.cfg:
```diff
@@ -1,5 +1,5 @@
 [flake8]
-exclude = build,dist,docs,.git,.tox,./two_factor/migrations/*.py,./two_factor/views/__init__.py,./tests/migrations/*
+exclude = build,dist,docs,.git,.tox,./two_factor/migrations/*.py,./two_factor/plugins/*/migrations/*.py,./two_factor/views/__init__.py,./tests/migrations/*
```
Didn't feel like changing current behaviour which is to have flake8 ignore migrations. Perhaps this one merits its own PR.

claudep replied:
Absolutely, this would be for another PR (you begin to know me!)

jpaniagualaconich replied:
i just made my only migration flake8-compatible to route around the controversy. Still think excluding migrations from flake8 checks is the way to go, though, as it's auto-generated code that's hardly ever modified by hand.

jpaniagualaconich commented on docs/installation.rst:
```diff
@@ -116,5 +116,43 @@ You could also do this using Django's `manage.py shell`:
+    INSTALLED_APPS = (
+        ...
+        'two_factor.plugins.webauthn',
+    )
```
All the other codeblocks in the same file were doing INSTALLED_APPS as a tuple. It's fine by me if INSTALLED_APPS is a list but in that case all of them should be, so that's what I did.

claudep replied:
Suggested as a separate PR in #538

jpaniagualaconich commented on tests/settings.py:
```diff
@@ -5,10 +5,18 @@
 SECRET_KEY = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890'
 
+ALLOWED_HOSTS = ['localhost']
+
```
not needed. I took it out.

jpaniagualaconich commented on tests/test_views_setup.py:
```diff
@@ -61,6 +66,48 @@ def test_setup_only_generator_available(self):
         self.assertEqual(1, self.user.totpdevice_set.count())
 
+    @skipUnless(webauthn, 'package webauthn is not present')
+    def test_setup_webauthn(self):
```
yes, that's where it should live now. Moved.

jpaniagualaconich commented on two_factor/plugins/webauthn/__init__.py:
```diff
@@ -0,0 +1,4 @@
+import django
+
+if django.VERSION <= (3, 2):
+    default_app_config = 'two_factor.plugins.webauthn.apps.TwoFactorWebauthnConfig'
```
gone.

jpaniagualaconich commented on two_factor/plugins/webauthn/admin.py:
```diff
@@ -0,0 +1,11 @@
+
+class WebauthnDeviceAdmin(admin.ModelAdmin):
+    list_display = ('user', 'name', 'created_at', 'last_used_at', 'confirmed')
+    raw_id_fields = ('user',)
```
done.

jpaniagualaconich commented on two_factor/plugins/webauthn/admin.py:
```diff
@@ -0,0 +1,11 @@
+    raw_id_fields = ('user',)
+
+
+admin.site.register(WebauthnDevice, WebauthnDeviceAdmin)
```
done.

jpaniagualaconich commented on two_factor/plugins/webauthn/apps.py:
```diff
@@ -0,0 +1,43 @@
+        except ImportError:
+            webauthn = None
+
+        if not webauthn:
```
you're right. done.

jpaniagualaconich commented on two_factor/plugins/webauthn/views.py:
```diff
@@ -0,0 +1,45 @@
+            raise Http404()
+
+        context = super().get_context_data(*args, **kwargs)
+        context['debug'] = settings.DEBUG
```
ok

jpaniagualaconich commented on two_factor/plugins/webauthn/templates/two_factor_webauthn/create_credential.js:
```diff
@@ -0,0 +1,37 @@
+    form.submit();
+
+}, (reason) => {
+    {% if debug %}console.error("Registration error: ", reason);{% endif %}
```
ok

jpaniagualaconich commented on two_factor/plugins/webauthn/tests/test_e2e.py:
```diff
@@ -0,0 +1,114 @@
+    timeout = 8
+
+    def setUp(self):
+        self.base_url = 'http://{}:{}'.format(self.host, self.port)
```
ok

jpaniagualaconich commented on two_factor/utils.py:
```diff
@@ -57,7 +55,3 @@ def totp_digits():
-
-
-def get_available_methods():
-    return [(m.code, m.verbose_name) for m in registry.get_methods()]
```
ok, it's on #539 now.

jpaniagualaconich commented on two_factor/views/core.py:
```diff
@@ -206,16 +208,17 @@ def get_form_kwargs(self, step=None):
+            kwargs['initial_device'] = self.get_device(step)
+        if 'request' in form_params:
+            kwargs['request'] = self.request
+        return kwargs
```
ok: #541

jpaniagualaconich commented on two_factor/views/core.py:
```diff
@@ -536,16 +543,19 @@ def done(self, form_list, **kwargs):
+        if 'device' in form_params:
+            kwargs['device'] = self.get_device()
+        if 'request' in form_params:
+            kwargs['request'] = self.request
```
ok: #541

jpaniagualaconich commented on two_factor/plugins/webauthn/forms.py:
```diff
@@ -0,0 +1,140 @@
+
+            new_sign_count = verify_authentication_response(
+                device.public_key, device.sign_count, self.webauthn_rp, self.webauthn_origin, challenge, token)
+        except Exception as exc:
```
ok, made it more specific.

jpaniagualaconich commented on two_factor/plugins/webauthn/forms.py:
```diff
@@ -0,0 +1,140 @@
+
+        try:
+            RegistrationCredential.parse_raw(token)
+        except Exception as exc:
```
ok, made it more specific.

jpaniagualaconich commented on two_factor/templates/two_factor/core/login.html:
```diff
@@ -41,4 +41,6 @@ <h1>{% block title %}{% trans "Login" %}{% endblock %}</h1>
     {% include "two_factor/_wizard_actions.html" %}
   </form>
+
+  {{ form.media }}
```
done.

jpaniagualaconich commented on CHANGELOG.md:
```diff
@@ -43,6 +43,8 @@
 ### Added
 - Translations for new languages: Hausa, Japanese, Vietnamese
 - Django 4.0 support
+- WebAuthn support
```
right, done.

claudep commented on docs/configuration.rst:
```diff
@@ -160,6 +160,30 @@ Fake Gateway
 
+WebAuthn Settings
+-----------------
+
```
Wow, this is a nice and welcome improvement, great!

claudep commented:
I guess we are approaching the merge moment, eventually! What do you think about rebasing and squashing the commits? Or we can do it during the merge, too.

jpaniagualaconich commented:
@claudep great idea. done.

claudep commented:
@moggers87, I think this is ready to go. can you give a last review or your approval?

claudep commented:
Thanks a LOT @jpaniagualaconich for the constant rebasing and great patience to add this new functionality.

jpaniagualaconich commented:
woohoo! 
thanks @claudep @moggers87 and all.

Frikster commented:
Been lurking and waiting for this to merge for over a year, haha. Congrats on the merge.

Be-ing commented:
Thank you @jpaniagualaconich, @moggers87, and @claudep!

Crocmagnon commented:
🎉  That's great news! When can we expect a release including this feature?
